tests/unit/test_charm.py

```diff
@@ -373,60 +404,73 @@ def test_install_apt_packages_add_package_failure(self, update, add_package):
+        self.harness.charm._add_mongodb_org_repository()
 
+        # verify we don't add repo when an exception occurs and that we enter blocked state
+        from_repo_line.assert_not_called()
+        import_key.assert_not_called()
+        self.assertEqual(repo_map.return_value, set())
         self.assertEqual(self.harness.charm.unit.status, BlockedStatus("couldn't install MongoDB"))
 
     @patch("charm.apt.RepositoryMapping")
+    @patch("charm.apt.DebianRepository.import_key")
     @patch("charm.apt.DebianRepository.from_repo_line")
     @patch("charm.urlopen")
     def test_add_mongodb_org_repository_cant_create_list_file_blocks(
-        self, urlopen, from_repo_line, get_mapping
+        self, urlopen, from_repo_line, import_key, repo_map
     ):
-        get_mapping.return_value = {}
+        repo_map.return_value = set()
         exceptions = [
             apt.InvalidSourceError("invalid source message"),
             ValueError("value message"),
         ]
-        log_messages = [
-            "ERROR:charm:failed to add repository, invalid source: invalid source message",
-            "ERROR:charm:failed to add repository: value message",
-        ]
-        for exception, log_message in zip(exceptions, log_messages):
-            with self.assertLogs("charm", "ERROR") as logs:
-                from_repo_line.side_effect = exception
-                self.harness.charm._add_mongodb_org_repository()
-                self.assertIn(log_message, "".join(logs.output))
+        for exception in exceptions:
+            from_repo_line.side_effect = exception
+            self.harness.charm._add_mongodb_org_repository()
 
+            # verify we don't add repo when an exception occurs and that we enter blocked state
+            import_key.assert_not_called()
+            self.assertEqual(repo_map.return_value, set())
             self.assertEqual(
                 self.harness.charm.unit.status, BlockedStatus("couldn't install MongoDB")
             )
 
-    @patch("charm.apt.RepositoryMapping")
+    # Will need to update this once I figure out how to mock RepositoryMapping properly
     @patch("charm.apt.DebianRepository.import_key")
     @patch("charm.apt.DebianRepository.from_repo_line")
-    @patch("charm.urlopen")
+    @patch("charm.apt.RepositoryMapping")
     def test_add_mongodb_org_repository_already_added_skips(
-        self, urlopen, from_repo_line, import_key, repo_map
+        self, repo_map, from_repo_line, import_key
     ):
-        repo_map.return_value = ["deb-https://repo.mongodb.org/apt/ubuntu-focal/mongodb-org/5.0"]
+        repo_map.return_value = {"deb-https://repo.mongodb.org/apt/ubuntu-focal/mongodb-org/5.0"}
+
+        # verify we don't add repo when we already have repo
         self.harness.charm._add_mongodb_org_repository()
-        urlopen.assert_called()
         from_repo_line.assert_not_called()
+        import_key.assert_not_called()
+        self.assertEqual(
```

Comment:
This assertion is totally unnecessary because we have mocked the value in this very method and we have simply repeated the string twice in the space of fewer than 10 lines of code.

This also made me think that, perhaps, a better way to implement _add_mongodb_org_repository() would be to parameterise what we are adding.  If we did this, then the unit test itself would be less brittle as it would not have to repeat exactly the same strings as used in the implementation.

Reply:
I think its good to mock this function since we don't want to import gpg keys for each run of the unit test. This would mean that if a developer contributes to this project and runs the unit test they will be importing the key onto their machine.

But we do want to check that `_add_mongodb_org_repository` calls import key. So we should in fact `import_key.assert_not_called()`

Reply:
As for `_add_mongodb_org_repository()` what exactly would you want parameterised? I am open to changing this function and want to be on the same page as you. 😄  (although things like the URL, repo_name, and repo_line) are constant since this charm is MongoDB vs5

Reply:
This case is the same as before.  Most of the objects are temporary and only exist in memory.  They are needed for the following call to work:
`                repositories.add(repo)`
This is the real one that would cause a change in the system but, in this case, you have mocked it.  So, all we need is to assert that this add was not called in this test.  If this is still not clear, let's talk about it.

As for the parameterisation, it is not strictly mandatory for now but we discussed the advantages when thinking about the future with both Mongo and Percona and perhaps more than one version for each.